core/src/main/java/org/elasticsearch/percolator/TopMatchingPercolatorType.java

```diff
@@ -0,0 +1,187 @@
+                    finalMatches.add(new PercolateResponse.Match(index, match, score));
+                }
+            }
+        } else {
+            int[] slots = new int[shardResults.size()];
+            while (true) {
+                float lowestScore = Float.NEGATIVE_INFINITY;
+                int requestIndex = -1;
+                int itemIndex = -1;
+                for (int i = 0; i < shardResults.size(); i++) {
+                    int scoreIndex = slots[i];
+                    float[] scores = shardResults.get(i).scores();
+                    if (scoreIndex >= scores.length) {
+                        continue;
+                    }
+
+                    float score = scores[scoreIndex];
+                    int cmp = Float.compare(lowestScore, score);
+                    // TODO: Maybe add a tie?
+                    if (cmp < 0) {
+                        requestIndex = i;
+                        itemIndex = scoreIndex;
+                        lowestScore = score;
+                    }
+                }
+
+                // This means the shard matches have been exhausted and we should bail
+                if (requestIndex == -1) {
+                    break;
+                }
+
+                slots[requestIndex]++;
+
+                PercolateShardResponse shardResponse = shardResults.get(requestIndex);
+                Text index = new StringText(shardResponse.getIndex());
+                Text match = new BytesText(new BytesArray(shardResponse.matches()[itemIndex]));
+                float score = shardResponse.scores()[itemIndex];
+                if (!shardResponse.hls().isEmpty()) {
+                    Map<String, HighlightField> hl = shardResponse.hls().get(itemIndex);
+                    finalMatches.add(new PercolateResponse.Match(index, match, score, hl));
+                } else {
+                    finalMatches.add(new PercolateResponse.Match(index, match, score));
+                }
+                if (finalMatches.size() == requestedSize) {
+                    break;
+                }
+            }
+        }
+
+        assert !shardResults.isEmpty();
+        InternalAggregations reducedAggregations = reduceAggregations(shardResults, headersContext);
+        return new PercolatorService.ReduceResult(foundMatches, finalMatches.toArray(new PercolateResponse.Match[finalMatches.size()]), reducedAggregations);
+    }
+
+    @Override
+    TopScoreDocCollector getCollector(int size) {
+        return TopScoreDocCollector.create(size);
+    }
+
+    @Override
+    PercolateShardResponse processResults(PercolateContext context, PercolatorQueriesRegistry registry, TopScoreDocCollector collector) throws IOException {
+        TopDocs topDocs = collector.topDocs();
+        long count = topDocs.totalHits;
+        List<BytesRef> matches = new ArrayList<>(topDocs.scoreDocs.length);
+        float[] scores = new float[topDocs.scoreDocs.length];
+        List<Map<String, HighlightField>> hls = null;
```

Comment:
could we initialize it to Collections.emptyList() and then check whether the list is empty instead of non-null so that we don't have this `if` statement to pick a collector at the end of the method?